### PR TITLE
V2 to v3 fragment

### LINKED ISSUE #10843: Convert v2 API test fragments to v3
While all programatically constructed test configs will be v3 (via API boosting rewriting), there are still v2 JSON/YAML fragments in test code. Many of these will "just work" as v3, but some may have v2 features that will have them treated as v2. This becomes confusing when they are extended by developers who want to model v3 features.

These need to be audited and updated as appropriate. Also, the names of methods such as `parseAccessLogFromV2Yaml` should be updated to reflect v3 when operating on fragments that should be v3.

This is somewhat similar to the work done by @dmitri-d for examples in https://github.com/envoyproxy/envoy/pull/10644 as part of https://github.com/envoyproxy/envoy/issues/9735.

CC @alyssawilk

### PR DESCRIPTION
**Pull request by ankatare**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message:  fragment changes from v2 to v3 API
Additional Description: 
Risk Level:Low
Testing: integration and format testing
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #10843 
[Optional Deprecated:]

### CODE REVIEW

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 1]
```diff
@@ -568,20 +578,21 @@ TEST_F(ClusterManagerImplTest, RingHashLoadBalancerInitialization) {
...
+                address: 127.0.0.1
+                port_value: 8000
+        - endpoint:
+            address:
+              socket_address:
+                address: 127.0.0.1
+                port_value: 8001
   )EOF";
-  create(parseBootstrapFromV2Yaml(yaml));
+  create(parseBootstrapFromV3Yaml(yaml));
```
Comment (htuch) [inline test/common/upstream/cluster_manager_impl_test.cc:595]: Have you update `parseBootstrapFromV3Yaml` to force v3?

File: test/common/upstream/utility.h [hunk 2]
```diff
@@ -124,7 +137,7 @@ makeLocalityWeights(std::initializer_list<uint32_t> locality_weights) {
 }
 
 inline envoy::config::core::v3::HealthCheck
-parseHealthCheckFromV2Yaml(const std::string& yaml_string) {
+parseHealthCheckFromV3Yaml(const std::string& yaml_string) {
   envoy::config::core::v3::HealthCheck health_check;
   TestUtility::loadFromYamlAndValidate(yaml_string, health_check);
```
Comment (htuch) [inline test/common/upstream/utility.h:142]: Do you need to force v3 here so we don't get any accidental v2 configs going through?

**Comment by ankatare** [comment]
@htuch Thanks for your review comments. I will update code as per your suggestion. Would be helpful if you could provide some code ref for the same.

**Comment by htuch** [comment]
@ankatare you want to invoke `MessageUtil::loadFromYaml` with do_boosting set to false.

**Comment by ankatare** [comment]
@htuch ok, I  will do that.  Thanks for Help !!!

**Comment by ankatare** [comment]
@htuch : please see possible code changes and it's implications.
 
**File:** test/test_common/utility.h
**Function Change:** 
static void loadFromYaml(const std::string& yaml, Protobuf::Message& message,
                           bool preserve_original_type = false) {
    MessageUtil::loadFromYaml(yaml, message, ProtobufMessage::getStrictValidationVisitor(),**false**);
    if (!preserve_original_type) {
      Config::VersionConverter::eraseOriginalTypeInformation(message);
    }
  }

Implications : Almost every test cases under test/.. will be affected due to this change and will be required some rework. 

Please suggest if it is good to go. .

**Comment by htuch** [comment]
@ankatare why did you go with type erasure vs. avoiding boosting in the first place?

**Comment by ankatare** [comment]
@htuch thanks for your quick comments. 
Do you means to set "avoid do_boosting" from inside
function : MessageUtil::loadFromYam () {}. as i see execution of actual logic is happening inside MessageUtil::loadFromJson () {}

**Comment by htuch** [comment]
@ankatare see https://github.com/envoyproxy/envoy/blob/582eb0ed1df4681063d551cb8966990599a2a5ba/source/common/protobuf/utility.h#L224, the ability to control `do_boosting` already exists. I think this should be plumbed through to the test wrappers.

File: test/extensions/health_checkers/redis/config_test.cc [hunk 3]
```diff
@@ -39,7 +39,7 @@ TEST(HealthCheckerFactoryTest, DEPRECATED_FEATURE_TEST(CreateRedisDeprecated)) {
   EXPECT_NE(
       nullptr,
       dynamic_cast<CustomRedisHealthChecker*>(
-          factory.createCustomHealthChecker(Upstream::parseHealthCheckFromV2Yaml(yaml), context)
+          factory.createCustomHealthChecker(Upstream::parseHealthCheckFromV3Yaml(yaml), context)
```
Comment (htuch) [inline test/extensions/health_checkers/redis/config_test.cc:42]: FYI, I don't think we want to change the parsing inside tests marked `DEPRECATED_FEATURE_TEST`. These should be v2 configs.

**Comment by ankatare** [comment]
@htuch Thanks for comments. I will get back as soon as possible

File: test/extensions/health_checkers/redis/config_test.cc (hunk 3 above)
Comment (ankatare) [reply test/extensions/health_checkers/redis/config_test.cc:42]: Ok. I will update it.

**Comment by ankatare** [comment]
@htuch Hi,  functionality of do_boosting is already implemented in https://github.com/envoyproxy/envoy/blob/master/source/common/protobuf/utility.h#L222.
In my opinion i just need to set value of **do_boosting to false** (to avoid boosting) here and use in the test codes ... seems  it was written to boost V2 config etc. Please let me know if my understanding is not correct here.

**Comment by htuch** [comment]
@ankatare yes, when you want to force it to parse as v2, you need to set `do_boosting` to false.

**Comment by ankatare** [comment]
@htuch while applying avoid-boosting ( do_boosting = false ) in test code, 2 test cases are failing. (remaining are ok)
it seems for both to get passed boosting is required. 
they are below test cases 
https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/cluster_manager_impl_test.cc#L419 
and
https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/cluster_manager_impl_test.cc#L539
please suggest way forward to proceed as we can't change exception output ?

**Comment by ankatare** [comment]
@htuch or suggest from whom to take help.

**Comment by htuch** [comment]
@ankatare are you trying to treat these examples as v2 or v3?

**Comment by ankatare** [comment]
@htuch seems, these test cases using v3 config only.

**Comment by htuch** [comment]
@ankatare and you are saying that when you parse them as v3, with `do_boosting = false`, this fails?

**Comment by ankatare** [comment]
@htuch yes.

**Comment by htuch** [comment]
You might want to run the test with `--test_arg="-l trace"` and see if there is some useful debug message. It should definitely be possible to parse v3 config as v3 without any boosting.

**Comment by ankatare** [comment]
@htuch ok. Let me check. Will get back ASAP

**Comment by ankatare** [comment]
@htuch see debug logs for
 https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/cluster_manager_impl_test.cc#L419
test/common/upstream/cluster_manager_impl_test.cc:397: Failure
Value of: std::string(e.what())
Expected: is equal to 0x25cf98 pointing to "cluster: LB policy hidden_envoy_deprecated_ORIGINAL_DST_LB is not valid for Cluster type STATIC. 'ORIGINAL_DST_LB' is allowed only with cluster type 'ORIGINAL_DST'"
  Actual: "Protobuf message (type envoy.config.bootstrap.v3.Bootstrap reason INVALID_ARGUMENT:(static_resources.clusters[0].lb_policy): invalid value \"original_dst_lb\" for type TYPE_ENUM) has unknown fields"
Stack trace:
  0x15b9dd1: Envoy::Upstream::(anonymous namespace)::ClusterManagerImplTest_OriginalDstLbRestriction2_Test::TestBody()
  0x3816404: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x38079be: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x37f1c13: testing::Test::Run()
  0x37f27fd: testing::TestInfo::Run()
... Google Test internal frames ...

[  FAILED  ] ClusterManagerImplTest.OriginalDstLbRestriction2 (201 ms)

**Comment by ankatare** [comment]
@htuch following code is executing. i checked with debug output
https://github.com/envoyproxy/envoy/blob/master/source/common/protobuf/utility.cc#L310

**Comment by htuch** [comment]
@ankatare that is v2 config if it's using ORIGINAL_DST_LB, which was deprecated. Should this be a fragment you hand convert to v3?

**Comment by ankatare** [comment]
@htuch yeah, while looking at this particular field it seems V2 config and it is deprecated as per doc https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster.proto#envoy-api-enum-cluster-lbpolicy.
and suggest to use "cluster_provided" as a value for lb_policy in V3. 
but i have not modified it, as this particular test code is meant for original_dst_lb only for type static. ( looking at exception o/p ) 
please guide me as  i am bit confused here :(

**Comment by htuch** [comment]
In this case, I think we should allow boosting, as the `EXPECT_THROW` is trying to show end-to-end what is expected. So, this is basically parse v2 with explicit boosting.

**Comment by ankatare** [comment]
@htuch Thanks a lot for this help. I will update code and get back ASAP.

File: test/common/upstream/cluster_manager_impl_test.cc (hunk 1 above)
Comment (ankatare) [reply test/common/upstream/cluster_manager_impl_test.cc:646]: @htuch yes. code changes is done as per comments.

File: test/common/upstream/utility.h (hunk 2 above)
Comment (ankatare) [reply test/common/upstream/utility.h:129]: @htuch yes

**Comment by ankatare** [comment]
@htuch code changes done and as per suggestion max of test cases would be forced for v3 except for few test cases where we need API boosting.

**Comment by ankatare** [comment]
@htuch this time i update force V3 for "parseBootstrapFromV3Yaml". please review it, if it is ok. i will start changing code for remaining "parseFOOFromV3Yaml". Moreover i missed to push changes for "DEPRECATED TEST CASES", will do it ASAP as per your comments.

File: test/test_common/utility.h [hunk 4]
```diff
@@ -533,6 +533,19 @@ class TestUtility {
...
+    if (avoid_boosting) {
+      MessageUtil::loadFromYaml(yaml, message, ProtobufMessage::getStrictValidationVisitor(),
+                                false);
+    } else {
+      MessageUtil::loadFromYaml(yaml, message, ProtobufMessage::getStrictValidationVisitor());
+    }
+    if (!preserve_original_type) {
+      Config::VersionConverter::eraseOriginalTypeInformation(message);
+    }
+  }
```
Comment (htuch) [inline test/test_common/utility.h:547]: You should define the `loadFromYaml` below in terms of this function.

File: test/common/upstream/utility.h (hunk 2 above)
Comment (htuch) [reply test/common/upstream/utility.h:129]: This one is still missing forcing v3.. can you ping me when this is ready for another pass? Thanks.
/wait

**Comment by ankatare** [comment]
@htuch i have incorporated all of your comments on force V3. Thanks.

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 5]
```diff
@@ -157,7 +158,7 @@ TEST_F(ClusterManagerImplTest, MultipleProtocolClusterFail) {
       http_protocol_options: {}
   )EOF";
   EXPECT_THROW_WITH_MESSAGE(
-      create(parseBootstrapFromV3Yaml(yaml)), EnvoyException,
+      create(parseBootstrapFromV3Yaml(yaml, false)), EnvoyException,
```
Comment (htuch) [inline test/common/upstream/cluster_manager_impl_test.cc:161]: Since this isn't a deprecated test, how come we don't avoid boosting? Ideally we update all tests to v3 that aren't deprecated.

File: test/common/upstream/health_checker_impl_test.cc [hunk 6]
```diff
@@ -167,7 +167,7 @@ class HttpHealthCheckerImplTest : public testing::Test, public HealthCheckerTest
       codec_client_type: Http2
     )EOF";
 
-    allocHealthChecker(yaml);
+    allocHealthChecker(yaml, true);
```
Comment (htuch) [inline test/common/upstream/health_checker_impl_test.cc:170]: Why  isn't `avoid_boosting` default true here, wouldn't that be less churn?

File: test/common/upstream/utility.h [hunk 7]
```diff
@@ -123,6 +123,14 @@ makeLocalityWeights(std::initializer_list<uint32_t> locality_weights) {
   return std::make_shared<LocalityWeights>(locality_weights);
 }
 
+inline envoy::config::core::v3::HealthCheck
+parseHealthCheckFromV3Yaml(const std::string& yaml_string, bool avoid_boosting = false) {
```
Comment (htuch) [inline test/common/upstream/utility.h:127]: I would think we would want to avoid boosting by default, and only in a few cases, e.g. deprecated tests, allow boosting.

**Comment by ankatare** [comment]
@htuch Thanks for comments. i will get back ASAP

**Comment by ankatare** [comment]
@htuch need help to understand this build error , things were fine in my local env. i posted this issue in envoy-ci channel as well.

**Comment by ankatare** [comment]
@htuch please review changes. i incorporated all of your comments.

File: test/common/tcp_proxy/tcp_proxy_test.cc [hunk 8]
```diff
@@ -53,9 +53,10 @@ using ::Envoy::Network::UpstreamServerName;
 
 namespace {
 Config constructConfigFromYaml(const std::string& yaml,
-                               Server::Configuration::FactoryContext& context) {
+                               Server::Configuration::FactoryContext& context,
+                               bool avoid_boosting = true) {
   envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy tcp_proxy;
-  TestUtility::loadFromYamlAndValidate(yaml, tcp_proxy);
+  TestUtility::loadFromYamlAndValidate(yaml, tcp_proxy, false, avoid_boosting);
```
Comment (htuch) [inline test/common/tcp_proxy/tcp_proxy_test.cc:59]: Isn't the third param to `loadFromYamlAndValidate` a validation visitor?

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 9]
```diff
@@ -442,7 +443,7 @@ TEST_F(ClusterManagerImplTest, OriginalDstLbRestriction2) {
                 port_value: 11001
   )EOF";
 
-  EXPECT_THROW_WITH_MESSAGE(create(parseBootstrapFromV3Yaml(yaml)), EnvoyException,
+  EXPECT_THROW_WITH_MESSAGE(create(parseBootstrapFromV3Yaml(yaml, false)), EnvoyException,
```
Comment (htuch) [inline test/common/upstream/cluster_manager_impl_test.cc:446]: Since this isn't a deprecated feature test, why avoid boosting?

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 10]
```diff
@@ -559,7 +560,7 @@ TEST_F(ClusterManagerImplTest, SubsetLoadBalancerOriginalDstRestriction) {
         - keys: [ "x" ]
   )EOF";
 
-  EXPECT_THROW_WITH_MESSAGE(create(parseBootstrapFromV3Yaml(yaml)), EnvoyException,
+  EXPECT_THROW_WITH_MESSAGE(create(parseBootstrapFromV3Yaml(yaml, false)), EnvoyException,
```
Comment (htuch) [inline test/common/upstream/cluster_manager_impl_test.cc:563]: Ditto

File: test/extensions/filters/network/http_connection_manager/config_test.cc [hunk 11]
```diff
@@ -40,10 +40,10 @@ namespace NetworkFilters {
 namespace HttpConnectionManager {
 
 envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager
-parseHttpConnectionManagerFromV2Yaml(const std::string& yaml) {
+parseHttpConnectionManagerFromV2Yaml(const std::string& yaml, bool avoid_boosting = true) {
```
Comment (htuch) [inline test/extensions/filters/network/http_connection_manager/config_test.cc:43]: v2 or v3 YAML?

File: test/extensions/filters/network/http_connection_manager/config_test.cc [hunk 12]
```diff
@@ -379,8 +379,8 @@ stat_prefix: router
   EXPECT_CALL(http_tracer_manager_, getOrCreateHttpTracer(Pointee(ProtoEq(inlined_tracing_config))))
       .WillOnce(Return(http_tracer_));
 
-  HttpConnectionManagerConfig config(parseHttpConnectionManagerFromV2Yaml(yaml_string), context_,
-                                     date_provider_, route_config_provider_manager_,
+  HttpConnectionManagerConfig config(parseHttpConnectionManagerFromV2Yaml(yaml_string, false),
```
Comment (htuch) [inline test/extensions/filters/network/http_connection_manager/config_test.cc:382]: Again, how come we are avoiding boosting for configs that presumably should be updated to v3?

**Comment by ankatare** [comment]
@htuch for some of the test cases i need to allow do_boosting (even they are not DEPRECATED)  as the EXPECT_THROW is trying to show end-to-end what is expected. So, this is basically parse v2 with explicit boosting. **same we discussed above.**

**Comment by ankatare** [comment]
@htuch for this comments i would say yes, it is a default parameter as false for loadFromYamlAndValidate  . Since i needed to parse avoid_boosting as 4th parameter in function argument. 

image

File: test/extensions/filters/network/http_connection_manager/config_test.cc (hunk 12 above)
Comment (htuch) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:382]: This one I still have a question with.

File: test/test_common/utility.h [hunk 13]
```diff
@@ -534,8 +534,13 @@ class TestUtility {
   }
 
   static void loadFromYaml(const std::string& yaml, Protobuf::Message& message,
-                           bool preserve_original_type = false) {
-    MessageUtil::loadFromYaml(yaml, message, ProtobufMessage::getStrictValidationVisitor());
+                           bool preserve_original_type = false, bool avoid_boosting = false) {
+    if (avoid_boosting) {
+      MessageUtil::loadFromYaml(yaml, message, ProtobufMessage::getStrictValidationVisitor(),
+                                false);
```
Comment (htuch) [inline test/test_common/utility.h:540]: Why can't this be a single `loadFromYaml` for both cases with the last param `!avoid_boosting`?

**Comment by ankatare** [comment]
@htuch for this case,  to get it pass i just modified function arguments llist as in the another test cases i  modified testutil::loadFromYamlAndValidate and it was calling it internally. so i am planning to change name one by one. after code is merged  please suggest. 
image

**Comment by ankatare** [comment]
@htuch thanks a lot for all your time on review my code.

**Comment by htuch** [comment]
@ankatare no worries; where possible, can you reply using GitHub threads BTW? It's hard to follow conversations with screenshot (although my Slack conversations are fair game for screenshotting ;-) )

**Comment by ankatare** [comment]
@htuch really sorry and **it is my bad**. somehow i am not seeing reply on your individual inputs  in my window. it's only a button of "resolve conversation". 
i don't know how to enable option, please help.

**Comment by adisuissa** [comment]
> Looks good. One thing I'm curious about is we see almost no changes to the config; does this mean our config examples are v3 safe for the most part? @adisuissa was this basically what you also observed?

Most configs were either v3 or v2+upgrade safe.

File: test/test_common/utility.h [hunk 14]
```diff
@@ -556,9 +557,15 @@ class TestUtility {
 
   template <class MessageType>
   static void loadFromYamlAndValidate(const std::string& yaml, MessageType& message,
-                                      bool preserve_original_type = false) {
-    MessageUtil::loadFromYamlAndValidate(yaml, message,
-                                         ProtobufMessage::getStrictValidationVisitor());
+                                      bool preserve_original_type = false,
+                                      bool avoid_boosting = true) {
+    if (avoid_boosting) {
```
Comment (adisuissa) [inline test/test_common/utility.h:562]: nit: I think this new code block can become: `MessageUtil::loadFromYamlAndValidate(yaml, message, ProtobufMessage::getStrictValidationVisitor(), !avoid_boosting);`

File: source/common/protobuf/utility.h [hunk 15]
```diff
@@ -262,8 +262,13 @@ class MessageUtil {
 
   template <class MessageType>
   static void loadFromYamlAndValidate(const std::string& yaml, MessageType& message,
-                                      ProtobufMessage::ValidationVisitor& validation_visitor) {
-    loadFromYaml(yaml, message, validation_visitor);
+                                      ProtobufMessage::ValidationVisitor& validation_visitor,
+                                      bool avoid_boosting = true) {
+    if (avoid_boosting) {
```
Comment (adisuissa) [inline source/common/protobuf/utility.h:267]: nit: I think this if-else block can be changed to: `loadFromYaml(yaml, message, validation_visitor, !avoid_boosting);`

**Comment by ankatare** [comment]
@htuch @adisuissa : thanks for your review comments, i will get back ASAP.

**Comment by ankatare** [comment]
@htuch incorporated review comments. thanks for your time to review it.

**Comment by ankatare** [comment]
@adisuissa incorporated review comments. thanks for your time to review it.

File: test/extensions/filters/network/http_connection_manager/config_test.cc (hunk 12 above)
Comment (htuch) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:382]: Friendly ping on this one; why do we need to allow boosting here? This is the main thing left IMHO.
